### PR TITLE
Add support for building with the Meson build system

### PR DESCRIPTION
**Pull request by tp-m**
Hi,

Please find attached a set of patches that adds support for building fribidi with the [Meson build system](http://mesonbuild.com). 

I understand that from a maintainer's point of view these kind of unsolicited "add support for a different build system" patches are not always welcome, not least because they might add additional maintenance burden, so apologies in advance for that. I would be happy to help maintain them going forward, along with other GNOME folks.

The Meson build system is [being adopted](https://wiki.gnome.org/Initiatives/GnomeGoals/MesonPorting) by projects such as GNOME, GTK, GStreamer, and there is support for it in Pango as well.
    
It would be fantastic to have support for building fribidi with Meson in upstream fribidi, as it would facilitate builds of GTK, Pango, GStreamer etc. via Meson's subproject support. Fribidi is the only 'hard' dependency of Pango outside the GNOME stack, so this would help fill an important gap for builds outside of your usual Linux distro environment.

Unlike autotools, Meson also supports builds on Windows with MSVC, for those who care about this (Windows developers mostly, since it means they can use native Windows tools for debugging).

Cross-compilation is of course also supported, but I have not tested that yet. Meson provides out-of-the-box support for building native binaries and cross binaries in the same build, which seems something that is useful for fribidi.

To try the meson build:
```
$ meson --prefix=/tmp/prefix builddir
$ ninja -C builddir
$ ninja -C builddir test
$ ninja -C builddir install
```    
Build time comparison:
- autotools: 57 seconds
- meson: 10 seconds

I have set the Meson requirement to the latest stable Meson release, as I have not tested it with older releases. You can install a recent Meson version with pip3 if your system does not have a recent one. You can also run Meson straight from a git checkout, since it's just python code.

### CODE REVIEW

**Comment by tp-m** [comment]
Mmmh, don't think the travis failure is related to the patchset, since the penultimate commit passes fine, and the last one that triggers the failure just adds a bunch of file to EXTRA_DIST.

**Comment by khaledhosny** [comment]
I restarted the build in it passed (though it leaves me wondering why it failed in the first place, hopefullt not some hidden bug).

I think you need to rebase on master because I merged #59 which moved some files around, sorry for that.

File: doc/meson.build [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  'fribidi_remove_bidi_marks.3',
+  'fribidi_reorder_line.3',
+  'fribidi_reorder_nsm_status.3',
+  'fribidi_set_debug.3',
+  'fribidi_set_mirroring.3',
+  'fribidi_set_reorder_nsm.3',
+  'fribidi_shape.3',
+  'fribidi_shape_arabic.3',
+  'fribidi_shape_mirroring.3',
+  'fribidi_unicode_to_charset.3',
```
Comment (khaledhosny) [inline doc/meson.build]: This hard-codes man pages while autotools takes it from `lib/fribidi.def` and I don’t think we want this duplication.

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+#
+# If any functions have been added:
+#   interface_age = 0
+#   interface_version++
+#
+# If binary backwards compatibility has been broken:
+#   panic!
+
+interface_age = 0
+interface_version = 4
```
Comment (khaledhosny) [inline meson.build]: I don’t like the fact we will now have two places where we need to increment the version number and potentially one will be missed.

File: doc/meson.build (hunk 1 above)
Comment (tp-m) [reply doc/meson.build]: Fixed.

File: meson.build (hunk 2 above)
Comment (tp-m) [reply meson.build]: I'm not sure what to do about the versions. I don't think there is an easy fix for this. Even if we read these two from the configure.ac (ew), the fribidi version in `project()` in the first line needs to be hard-coded there first thing.

We could derive a suitable `interface_age` and `interface_version` from your fribidi minor version and micro version, but that's going to be painful in autotools as it doesn't do simple maths. What we do in GStreamer is to use maths in the meson file and have the formula as a comment in `configure.ac`.

Another option would be to hook a script into 'make check' or 'make distcheck' that makes sure versions are consistent in `meson.build` and `configure.ac`.

I think most maintainers tend to remember that they support two build systems and fix it up in both, it's rarely been a problem :)

**Comment by tp-m** [comment]
Thanks for the quick review! I've now rebased on top of your changes, which make things a bit simpler.

File: meson.build (hunk 2 above)
Comment (khaledhosny) [reply meson.build]: > I think most maintainers tend to remember that they support two build systems and fix it up in both, it's rarely been a problem :)

You would bet :slightly_smiling_face: Anyway, I’ll differ the opinion here to @dov and @behdad as they are the actual maintainers.

Comment (behdad) [reply meson.build]: Hahaha. I never update two build systems... Anyway, @dov's call. I'm not maintainer anymore.

Comment (tp-m) [reply meson.build]: How about we simply sprinkle a few `dnl keep in sync with meson.build` in `configure.ac` then? :)

**Comment by dov** [comment]
Thanks! I just read up about mason as a result of this pull request, and it though I don't understand the details, it seems to me like a huge improvement to autotools, that I could never fully groked. If is good enough for the gnome stack, it is certainly good enough for fribidi.

**Pull request merged**